### PR TITLE
Dev/add theme picker

### PR DESCRIPTION
**Pull request by ARADDCC002**
image

image

### CODE REVIEW

File: frontend/src/settings/display/ExampleDisplay.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import { patchCurrentUserSettings, useGetCurrentUserSettings } from 'actions/user'
+import { ThemeKey } from 'src/theme'
+
+interface ExampleDisplayProps {
+  theme: ThemeKey
+}
+export default function ExampleDisplay({ theme }: ExampleDisplayProps) {
+  const { mutateUserSettings } = useGetCurrentUserSettings()
+
+  async function setTheme() {
```
Comment (ARADDCC012) [inline frontend/src/settings/display/ExampleDisplay.tsx:12]: I'd prefer to keep the `setSomething` naming convention for updating react state vars. Consider renaming this function to something like `handleUpdateTheme`

File: frontend/src/settings/display/ExampleDisplay.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    const updatedSettings = await patchCurrentUserSettings({ theme: theme.key })
+    if (updatedSettings.status === 200) {
```
Comment (ARADDCC012) [inline frontend/src/settings/display/ExampleDisplay.tsx:14]: No error handling

File: frontend/src/settings/display/ExampleDisplay.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      mutateUserSettings()
+    }
+  }
+
+  return (
+    <ThemeProvider theme={theme}>
```
Comment (ARADDCC012) [inline frontend/src/settings/display/ExampleDisplay.tsx:20]: This should be able to be removed as it's already included in `_app.tsx`

File: frontend/src/theme.ts [hunk 4]
```diff
@@ -38,6 +38,11 @@ const defaultComponentOverrides: ThemeOptions['components'] = {
   },
 }
 
+export interface ThemeKey {
```
Comment (ARADDCC012) [inline frontend/src/theme.ts:41]: This is more of a mapping than a key, consider renaming to ThemeMapping

File: frontend/src/settings/display/ExampleDisplay.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { Box, Button } from '@mui/material'
+import { ThemeProvider } from '@mui/material/styles'
+import { patchCurrentUserSettings, useGetCurrentUserSettings } from 'actions/user'
+import { ThemeKey } from 'src/theme'
+
+interface ExampleDisplayProps {
+  theme: ThemeKey
```
Comment (ARADDCC012) [inline frontend/src/settings/display/ExampleDisplay.tsx:7]: This prop name is confusing as it's not a `Theme`. Consider renaming to `themeMapping` in line with one of my other comments.

File: frontend/types/types.ts [hunk 6]
```diff
@@ -373,3 +373,8 @@ export type ReviewRequestInterface = {
   createdAt: string
   updatedAt: string
 } & PartialReviewRequestInterface
+
+export interface UserSettings {
+  dn: string
+  theme: string
```
Comment (ARADDCC012) [inline frontend/types/types.ts:379]: It's hard to keep track of what `theme` means throughout this PR. Can we clarify that this is a `themeKey`? Sorry if this results in wide sweeping changes, but being able to tell that this is this is linked to the `key` will help with readability.

File: frontend/src/theme.ts [hunk 7]
```diff
@@ -234,3 +239,21 @@ export const darkTheme = createTheme({
     },
   },
 })
+
+export const ThemeNames = {
```
Comment (ARADDCC012) [inline frontend/src/theme.ts:243]: Generally the naming convention for enums is to use singular names, not plural. The reason being is that when you use the enum in code, you're selecting one thing from the choices available.

I think the same rules should apply here as it's our own way of using enums without all of the TS headache.

File: frontend/src/theme.ts [hunk 8]
```diff
@@ -234,3 +239,21 @@ export const darkTheme = createTheme({
...
+  Light: 'light',
+  Dark: 'dark',
+} as const
+
+const lightThemeKey = {
```
Comment (ARADDCC012) [inline frontend/src/theme.ts:248]: If you agree with my other comment, rename this to `lightThemeMapping`

File: frontend/src/theme.ts [hunk 9]
```diff
@@ -234,3 +239,21 @@ export const darkTheme = createTheme({
...
+  key: ThemeNames.Light,
+  theme: lightTheme,
+  title: 'Light',
+}
+const darkThemeKey = {
```
Comment (ARADDCC012) [inline frontend/src/theme.ts:253]: If you agree with my other comment, rename this to `darkThemeMapping`

**Review by ARADDCC012 — CHANGES_REQUESTED** [review]

File: frontend/actions/user.ts [hunk 10]
```diff
@@ -91,3 +99,26 @@ export function deleteUserToken(accessKey: TokenInterface['accessKey']) {
...
+interface UserSettingsResponse {
+  settings: UserSettings
+}
+
+export function useGetCurrentUserSettings() {
+  const { data, error, mutate } = useSWR<UserSettingsResponse, ErrorInfo>('/api/v2/user/settings', fetcher)
+
+  return {
+    mutateUserSettings: mutate,
+    userSettings: data?.settings || undefined,
```
Comment (ARADDCC012) [inline frontend/actions/user.ts:112]: Can be simplified to: `userSettings: data?.settings`

File: frontend/src/settings/ProfileTab.tsx [hunk 11]
```diff
@@ -8,9 +7,11 @@ type ProfileTabProps = {
 
 export default function ProfileTab({ user }: ProfileTabProps) {
   return (
-    <Box sx={{ px: 2, py: 4 }}>
-      <Typography fontWeight='bold'>Name</Typography>
-      <Typography>{user.dn}</Typography>
-    </Box>
+    <Container maxWidth='md'>
+      <Box>
```
[the remark this answers is not in the document]
Comment (ARADDCC002) [reply frontend/src/settings/ProfileTab.tsx:11]: This keeps the content centered to be consistent with the other tabs

File: frontend/src/settings/display/ExampleDisplay.tsx (hunk 3 above)
Comment (ARADDCC002) [reply frontend/src/settings/display/ExampleDisplay.tsx:20]: Each example display has its own theme (so you can see what the colours are!).

image

Ideally I'll improve it so they're not just buttons

**Pull request closed without merging**